File: cracking_coding/3_Stack_Queue/stack.py
from sys import maxsize
class Stack():
    def __init__(self, limit):
        self.top = 0
        self.stack = []

    def push(self, x):
        self.stack.append(x)
        self.top += 1
        return True

    def pop(self):
        if(self.isEmpty()):
            return str(-maxsize - 1) # return negative infinity

        return self.stack.pop()

    def isEmpty(self):
        return len(self.stack) == 0

s = Stack(10)
s.push(10)
s.push(20)
s.push(30)

print(s.pop())
print(s.pop())

File: cracking_coding/3_Stack_Queue/stack.cpp
#include <iostream>
#include <string>

using namespace std;

const int MAX = 1000;

class Stack{
    int top;
public:
    int a[MAX];

    Stack() { top= -1 ;}
    bool push(int x){
        if(top > MAX){
            cout << "Stack overflow" << endl;
            return false;
        }else{
            a[++top] = x;
            return true;
        }
    }
    int pop(){
        if(top < 0){
            cout << "Stack underflow" << endl;
            return 0;
        }
        else{
            int x = a[top--];
            return x;
        }
    }
    bool isEmpty(){
        return (top < 0);
    }
};

int main(){
    struct Stack s;
    s.push(10);
    s.push(20);
    s.push(30);
    cout << s.pop() << " popped" << endl;
    return 0;
}
